fe1-web/src/features/popcha/screens/__tests__/PopchaScanner.test.tsx

```diff
@@ -0,0 +1,221 @@
+      const url = new URL(mockUrl.toString());
+      url.searchParams.delete('scope');
+      await testInvalidUrl(url.toString());
+    });
+
+    it('shows error message with url with invalid scope', async () => {
+      const url = new URL(mockUrl.toString());
+      url.searchParams.set('scope', 'invalid');
+      await testInvalidUrl(url.toString());
+    });
+
+    it('shows error message with url with invalid response_type', async () => {
+      const url = new URL(mockUrl.toString());
+      url.searchParams.set('response_type', 'scope invalid response type');
+      await testInvalidUrl(url.toString());
+    });
+
+    it('shows error message with url with invalid response_mode', async () => {
+      const url = new URL(mockUrl.toString());
+      url.searchParams.set('response_mode', 'invalid response mode');
+      await testInvalidUrl(url.toString());
+    });
+
+    it('shows error message when login_hint does not match current laoId', async () => {
+      const url = new URL(mockUrl.toString());
+      url.searchParams.set('login_hint', 'invalid login hint');
+      await testInvalidUrl(url.toString());
+    });
+  });
+
+  describe('valid url sends correct response', () => {
+    it('sends correct response with valid url', async () => {
+      const { getByTestId } = render(
+        <FeatureContext.Provider value={contextValue}>
+          <PopchaScanner />
+        </FeatureContext.Provider>,
+      );
+
+      const scannerButton = getByTestId('popcha_scanner_button');
+      fireEvent.press(scannerButton);
+      fireScan(mockUrl.toString());
+      await waitFor(() => expect(mockToastShow).toHaveBeenCalledTimes(0));
+    });
+
+    it('does not show error message with valid response mode', async () => {
+      const url = new URL(mockUrl.toString());
+      url.searchParams.set('response_mode', 'query');
+      const { getByTestId } = render(
+        <FeatureContext.Provider value={contextValue}>
+          <PopchaScanner />
+        </FeatureContext.Provider>,
+      );
+
+      const scannerButton = getByTestId('popcha_scanner_button');
+      fireEvent.press(scannerButton);
+      fireScan(url.toString());
+      await waitFor(() => expect(mockToastShow).toHaveBeenCalledTimes(0));
+    });
+  });
+
+  describe('correct behavior when sending request', () => {
+    it('closes scanner when it is a successful request', async () => {
+      const url = new URL(mockUrl.toString());
+      (sendPopchaAuthRequest as jest.Mock).mockReturnValue(Promise.resolve());
+      const { getByTestId, getByText, toJSON } = render(
+        <FeatureContext.Provider value={contextValue}>
+          <PopchaScanner />
+        </FeatureContext.Provider>,
+      );
+      const scannerButton = getByTestId('popcha_scanner_button');
+      fireEvent.press(scannerButton);
+      fireScan(url.toString());
+      await waitFor(() => {
+        // wait for scanner to close
+        expect(getByText(STRINGS.popcha_open_scanner)).toBeTruthy();
```

Comment:
I used `getByText` because I want to wait for the "open scanner" to appear, I think a `getByTestId` wouldn't work since the button exists when the scanner is open (that was my initial issue with this test, I was only getting a snapshot test and it would take it with the scanner still open -> wouldn't wait for it to close)